### PR TITLE
feat: allow filename to be a `{Function}`

### LINKED ISSUE #143: feat: allow the option filename to be a function
Can we allow the filename and chunkFilename passed in trought the constructor to be functions instead of string only?

In our case it would make it easy to toggle the [hashcontent] for some bundles. 
We would pass the option as followed: 
`filename: (chunk)  =>  chunk.name === "NotHashedBundle" ? '[name].css' : '[name].[contenthash].css'`

Might be worth noting that the latest version of webpack also supports something simular to this in their output.
eg: `filename: (data) => data.chunk.name === "NotHashedBundle" ? '[name].js' : '[name].[chunkhash].js'`

### PR DESCRIPTION
**Pull request by berthertogen**
This PR contains a:

- [ ] **bugfix**
- [x] new **feature**
- [ ] **code refactor**
- [x] **test update** 
- [ ] **typo fix**
- [ ] **metadata update**

### Motivation / Use-Case

closes #143 

### Additional Info

Wasn't able to get the tests to run! But wrote a test for the filename-function.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -111,6 +111,11 @@ class MiniCssExtractPlugin {
       options
     );
     if (!this.options.chunkFilename) {
+      if (this.isFunction(this.options.filename)) {
+        throw new Error(
+          `You are required to provide a value for chunkFilename when using a Function for this.options.filename, please specify the [name], [id] or [chunkhash] in this function`
```
Comment (evilebottnawi) [inline src/index.js]: Remove this `throw`, it should be done after we integrate `schema` for options

File: src/index.js [hunk 2]
```diff
@@ -366,6 +374,17 @@ class MiniCssExtractPlugin {
     });
   }
 
+  getFilename(chunk, filename) {
+    return this.isFunction(filename) ? filename(chunk) : filename;
+  }
+
+  isFunction(functionToCheck) {
```
Comment (evilebottnawi) [inline src/index.js]: Remove this and use `typeof something === 'function'`

**Comment by aaarichter** [comment]
looking forward to this :octocat:

**Comment by berthertogen** [comment]
Sorry it took so long, I remove the console logs and made sure the filename function is called only once.

**Comment by glen-84** [comment]
Is there enough information in `chunk` to figure out the path used to reference the CSS file/the path on disk? (see #92)

File: src/index.js [hunk 3]
```diff
@@ -170,7 +170,11 @@ class MiniCssExtractPlugin {
                   renderedModules,
                   compilation.runtimeTemplate.requestShortener
                 ),
-              filenameTemplate: this.options.filename,
+              filenameTemplate: this.getFilename(
+                chunk,
+                this.options.filename,
+                this.options.processedFilename
```
Comment (ooflorent) [inline src/index.js]: What is this option? There is no documentation about it and it is not used by the test cases.

File: src/index.js [hunk 4]
```diff
@@ -366,6 +380,19 @@ class MiniCssExtractPlugin {
     });
   }
 
+  getFilename(chunk, filename, processedFilename) {
+    if (!processedFilename) {
+      processedFilename = this.isFunction(filename)
```
Comment (ooflorent) [inline src/index.js]: Use `return` here. Reassigning parameters result in _deopt_.

Comment (ooflorent) [reply src/index.js]: Could you use `typeof` instead please?

**Comment by michael-ciniawsky** [comment]
Closing in favor of #225 

 Thx :)

**Pull request closed without merging**